Adds dynamic and multi-mesh support for ray caster

timkobiolka opened the pull request:
# Description

This PR extends the ray caster to support multiple meshes in `mesh_prim_paths`, enabling it to interact with complex environments. Additionally, it now supports dynamic meshes, allowing real-time updates to mesh positions during simulation.

This update is fully backward-compatible with existing single-mesh configuration.

A test file is included which shows that multiple and dynamic meshes are processed by the updated ray caster.

Fixes #1804 #1395 #1259 #553 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

Bonseok-Gu99 commented:
isaacLab4.5/source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py", line 297, in _update_buffers_impl
    final_distances[mask] = distance[mask]
IndexError: The shape of the mask [10, 187] at index 0 does not match the shape of the indexed tensor [1, 187] at index 0

there is an error. During training quadruped robot.

cornuco-pia reviewed (changes requested):
it gives an error if we're adding raycaster to observations and resetting an env. you have some mismatch in array shapes, i believe this on lines 292-297 fixes it:
```
mask = distance < final_distances[env_ids]
                final_distances[env_ids][mask] = distance[mask]
                final_hits[env_ids][mask] = hit[mask]

        # update the ray hit data for the specified environments
        self._data.ray_hits_w[env_ids] = final_hits[env_ids]
```

Mayankm96 commented:
Thanks for the MR. This is an interesting addition. However, using a for loop over all the meshes will be rather slow. Especially if we go for multiple meshes per environment -- two mesh per environment with 4096 environments will mean 8192 separate warp kernel calls. This is rather expensive as launching a kernel takes time.

Is it possible to make a new warp kernel that parallelizes this as well?

timkobiolka commented:
> it gives an error if we're adding raycaster to observations and resetting an env. you have some mismatch in array shapes, i believe this on lines 292-297 fixes it:
> 
> ```
> mask = distance < final_distances[env_ids]
>                 final_distances[env_ids][mask] = distance[mask]
>                 final_hits[env_ids][mask] = hit[mask]
> 
>         # update the ray hit data for the specified environments
>         self._data.ray_hits_w[env_ids] = final_hits[env_ids]
> ```

Thanks for catching that! I have applied the fix 😃

timkobiolka commented:
> Thanks for the MR. This is an interesting addition. However, using a for loop over all the meshes will be rather slow. Especially if we go for multiple meshes per environment -- two mesh per environment with 4096 environments will mean 8192 separate warp kernel calls. This is rather expensive as launching a kernel takes time.
> 
> Is it possible to make a new warp kernel that parallelizes this as well?

Thanks for reviewing the MR and for the feedback! I initially structured it with a loop over the meshes to keep the logic more straightforward. However, I agree that parallelizing the kernel calls for all meshes would be much more efficient, especially at scale with many environments. I’ll look into adding a new Warp kernel this week to handle multiple meshes.

timkobiolka commented:
I have now implemented a new Warp kernel (multi_mesh_raycast_kernel) to efficiently handle multiple meshes in parallel. The previous for-loop has been replaced with a single batched kernel that processes all rays against multiple meshes simultaneously. I have also added caching for mesh-to-environment mappings to reduce CPU overhead between frames. This should significantly improve the raycasting performance, especially for larger numbers of environments and meshes. Let me know if any further adjustments or improvements are needed.

bikcrum commented on source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py:
```diff
@@ -250,14 +260,132 @@ def _update_buffers_impl(self, env_ids: Sequence[int]):
+                indices = torch.arange(mesh_view.count, device=self._device)
+                # get the current world transforms for the mesh prim (for dynamic updates)
+                mesh_pos, mesh_quat = mesh_view.get_world_poses(indices)
+                base_points = torch.tensor(mesh_info["base_points"], device=self._device)
```
I had to convert this to a float32 to make it work. It threw an error in quat_apply(). It may be worth noting it.

bikcrum commented:
Thanks for the PR. 

It might also worth adding regex processing of mesh prims in the interactive scene in the file https://github.com/timkobiolka/IsaacLab/blob/b24925868229ea915da447d59e266426831581f1/source/isaaclab/isaaclab/scene/interactive_scene.py#L590
```
#existing code

if hasattr(asset_cfg, "prim_path"):
      asset_cfg.prim_path = asset_cfg.prim_path.format(ENV_REGEX_NS=self.env_regex_ns)

# added line

if hasattr(asset_cfg, "mesh_prim_paths"):
      asset_cfg.mesh_prim_paths = [path.format(ENV_REGEX_NS=self.env_regex_ns) for path in asset_cfg.mesh_prim_paths]
```
Now this can be used as

```
camera_cfg: RayCasterCameraCfg = RayCasterCameraCfg(
     ....
     mesh_prim_paths=["{ENV_REGEX_NS}/object"],
)
```

cornuco-pia commented:
this here, at line 272 in ray_caster.py, takes a really long time to compute - adding 2 lidars increased my simulation time by 60 times. could it be possible to parallelize this too?
<img width="453" alt="Screenshot 2025-03-21 at 14 48 16" src="https://github.com/user-attachments/assets/89ee16eb-f917-4971-875e-11151a803c1c" />

amatsukaze233 commented:
Why can't the objects I imported be ray-cast
```

 workpiece2= AssetBaseCfg(
        prim_path="{ENV_REGEX_NS}/workpiece2",
        spawn=sim_utils.UsdFileCfg(
            usd_path="C:/Users/Administrator/IsaacLab/source/standalone/mydemo/data/container.usd",
            scale=(0.2, 0.2, 0.2),
            visual_material=sim_utils.PreviewSurfaceCfg(diffuse_color=(0.2, 0.98, 0.5),metallic=0.2),
            ),
            init_state=AssetBaseCfg.InitialStateCfg(pos=(-1.4,1.1,0.0),rot=(-0.28422,0.0,0.95876,0.0)),       
    )

    workpiece= RigidObjectCfg(
        prim_path="{ENV_REGEX_NS}/workpiece",
        spawn=sim_utils.UsdFileCfg( 
            usd_path="C:/Users/Administrator/IsaacLab/source/standalone/mydemo/data/cabinet.usd",
            scale=(0.2, 0.2, 0.2),
            collision_props=sim_utils.CollisionPropertiesCfg(),
            visual_material=sim_utils.PreviewSurfaceCfg(diffuse_color=(0.85, 0.78, 0.5),metallic=0.2),
            rigid_props=sim_utils.RigidBodyPropertiesCfg(disable_gravity=True,
                rigid_body_enabled=True,
                kinematic_enabled=True,),
            ),
            init_state=RigidObjectCfg.InitialStateCfg(pos=(0.2,0.5,0.5),rot=(-0.28422,0.0,0.95876,0.0)),    
     
    )

  movingcuboid= RigidObjectCfg(
        prim_path="{ENV_REGEX_NS}/MovingCuboid",
        spawn=sim_utils.MeshCuboidCfg(
            size=(1.5, 1.5, 0.2),
            rigid_props=sim_utils.RigidBodyPropertiesCfg(
                disable_gravity=True,
                rigid_body_enabled=True,
                kinematic_enabled=True,
            ),
            mass_props=sim_utils.MassPropertiesCfg(mass=1.0),
            collision_props=sim_utils.CollisionPropertiesCfg(),
            visual_material=sim_utils.PreviewSurfaceCfg(diffuse_color=(0.0, 1.0, 0.0), metallic=0.2),
        ),
    )

 ray_caster = RayCasterCfg(
        prim_path="/World/envs/env_.*/GCR/tcp",
        update_period=1 / 30,
        offset=RayCasterCfg.OffsetCfg(pos=(0.0, 0.0, 0.0),rot=(0.707,0.0,-0.707,0.0)),#tcp z
        mesh_prim_paths=["/World/envs/env_.*/workpiece2","/World/defaultGroundPlane","/World/envs/env_.*/workpiece","{ENV_REGEX_NS}/MovingCuboid"],
        attach_yaw_only=False,
        pattern_cfg=patterns.LidarPatternCfg(
            channels=50, vertical_fov_range=(-45, 45), horizontal_fov_range=(-45, 45), horizontal_res=90/50
        ),
        debug_vis=True,
        max_distance = 500.0,)

```

![20250409175256](https://github.com/user-attachments/assets/214f18b9-609f-4bee-a213-56ea4f253461)

qin-yuhao commented:
![53fb80d3-dc66-4c88-9891-36405c6ccba8](https://github.com/user-attachments/assets/49e59b45-f496-4fa3-84a0-6a0fb25b6809)
![a0ecc285-f65e-4267-a744-8e6827c967e9](https://github.com/user-attachments/assets/a393f0c7-10d0-4dc7-9fd9-d126fff775aa)
When I add "/World/envs/env_0/Robot/RR_calf/collisions/RR_calf" to the mesh_prim_paths list, the LidarPattern radar data seems to show errors in scripts\demos\sensors\raycaster_sensor.py.        `# Copyright (c) 2022-2025, The Isaac Lab Project Developers.
# All rights reserved.
#
# SPDX-License-Identifier: BSD-3-Clause

import argparse
import numpy as np

from isaaclab.app import AppLauncher

# add argparse arguments
parser = argparse.ArgumentParser(description="Example on using the raycaster sensor.")
parser.add_argument("--num_envs", type=int, default=1, help="Number of environments to spawn.")
# append AppLauncher cli args
AppLauncher.add_app_launcher_args(parser)
# parse the arguments
args_cli = parser.parse_args()

# launch omniverse app
app_launcher = AppLauncher(args_cli)
simulation_app = app_launcher.app

"""Rest everything follows."""

import torch

import isaaclab.sim as sim_utils
from isaaclab.assets import AssetBaseCfg
from isaaclab.scene import InteractiveScene, InteractiveSceneCfg
from isaaclab.sensors.ray_caster import RayCasterCfg, patterns
from isaaclab.utils import configclass
from isaaclab.utils.assets import ISAAC_NUCLEUS_DIR

##
# Pre-defined configs
##
from isaaclab_assets.robots.anymal import ANYMAL_C_CFG  # isort: skip
from isaaclab_assets.robots.mit import Mit_CFG

@configclass
class RaycasterSensorSceneCfg(InteractiveSceneCfg):
    """Design the scene with sensors on the robot."""

    # ground plane
    ground = AssetBaseCfg(
        prim_path="/World/Ground",
        spawn=sim_utils.UsdFileCfg(
            usd_path=f"{ISAAC_NUCLEUS_DIR}/Environments/Terrains/rough_plane.usd",
            scale=(1, 1, 1),
        ),
    )

    # lights
    dome_light = AssetBaseCfg(
        prim_path="/World/Light", spawn=sim_utils.DomeLightCfg(intensity=3000.0, color=(0.75, 0.75, 0.75))
    )

    # robot
    #robot = ANYMAL_C_CFG.replace(prim_path="{ENV_REGEX_NS}/Robot")
    robot = Mit_CFG.replace(prim_path="{ENV_REGEX_NS}/Robot")
    ray_caster = RayCasterCfg(
        #prim_path="{ENV_REGEX_NS}/Robot/base/lidar_cage",
        prim_path="{ENV_REGEX_NS}/Robot/base",
        update_period=1 / 60,
        offset=RayCasterCfg.OffsetCfg(pos=(0, 0, 0.5),rot=(0, -0.7071, 0, 0.7071)),
        mesh_prim_paths=["/World/Ground","/World/envs/env_0/Robot/RR_calf/collisions/RR_calf"],
        attach_yaw_only=True,
        # pattern_cfg=patterns.LidarPatternCfg(
        #     channels=100, vertical_fov_range=[-90, 90], horizontal_fov_range=[-90, 90], horizontal_res=1.0
        # ),

        pattern_cfg=patterns.PinholeLidarPatternCfg(
        vertical_channels=32,           # 16线激光雷达
        horizontal_channels=32,        # 水平128个点
        vertical_fov_range=[-22.5, 22.5],
        horizontal_fov_range=[-22.5, 22.5],
        ),

        debug_vis=not args_cli.headless,
    )


def run_simulator(sim: sim_utils.SimulationContext, scene: InteractiveScene):
    """Run the simulator."""
    # Define simulation stepping
    sim_dt = sim.get_physics_dt()
    sim_time = 0.0
    count = 0

    triggered = True
    countdown = 42

    # Simulate physics
    while simulation_app.is_running():

        if count % 500 == 0:
            # reset counter
            count = 0
            # reset the scene entities
            # root state
            # we offset the root state by the origin since the states are written in simulation world frame
            # if this is not done, then the robots will be spawned at the (0, 0, 0) of the simulation world
            root_state = scene["robot"].data.default_root_state.clone()
            root_state[:, :3] += scene.env_origins
            scene["robot"].write_root_pose_to_sim(root_state[:, :7])
            scene["robot"].write_root_velocity_to_sim(root_state[:, 7:])
            # set joint positions with some noise
            joint_pos, joint_vel = (
                scene["robot"].data.default_joint_pos.clone(),
                scene["robot"].data.default_joint_vel.clone(),
            )
            joint_pos += torch.rand_like(joint_pos) * 0.1
            scene["robot"].write_joint_state_to_sim(joint_pos, joint_vel)
            # clear internal buffers
            scene.reset()
            print("[INFO]: Resetting robot state...")
        # Apply default actions to the robot
        # -- generate actions/commands
        targets = scene["robot"].data.default_joint_pos
        # -- apply action to the robot
        scene["robot"].set_joint_position_target(targets)
        # -- write data to sim
        scene.write_data_to_sim()
        # perform step
        sim.step()
        # update sim-time
        sim_time += sim_dt
        count += 1
        # update buffers
        scene.update(sim_dt)

        # print information from the sensors
        print("-------------------------------")
        print(scene["ray_caster"])
        print("Ray cast hit results: ", scene["ray_caster"].data.ray_hits_w)

        if not triggered:
            if countdown > 0:
                countdown -= 1
                continue
            data = scene["ray_caster"].data.ray_hits_w.cpu().numpy()
            np.save("cast_data.npy", data)
            triggered = True
        else:
            continue


def main():
    """Main function."""

    # Initialize the simulation context
    sim_cfg = sim_utils.SimulationCfg(dt=0.005, device=args_cli.device)
    sim = sim_utils.SimulationContext(sim_cfg)
    # Set main camera
    sim.set_camera_view(eye=[3.5, 3.5, 3.5], target=[0.0, 0.0, 0.0])
    # design scene
    scene_cfg = RaycasterSensorSceneCfg(num_envs=args_cli.num_envs, env_spacing=2.0)
    scene = InteractiveScene(scene_cfg)
    # Play the simulator
    sim.reset()
    # Now we are ready!
    print("[INFO]: Setup complete...")
    # Run the simulator
    run_simulator(sim, scene)


if __name__ == "__main__":
    # run the main function
    main()
    # close sim app
    simulation_app.close()
`

AlexanderAbernathy commented:
Have you tested the RayCasterCamera? I encountered data errors after replacing the core scripts.
Error:
[Error] [omni.physx.plugin] PhysX error: Failed to unload CUDA module data, returned 710., FILE /builds/omniverse/physics/physx/source/cudamanager/src/CudaContextManager.cpp, LINE 804

Jason-xy commented:
I was wondering if there’s any update on when it might be merged into the main branch. I’ve run into the same issue when testing the RayCasterCamera after updating the core scripts:

```
[Error] [omni.physx.plugin] PhysX error: Failed to unload CUDA module data, returned 710., FILE /builds/omniverse/physics/physx/source/cudamanager/src/CudaContextManager.cpp, LINE 804
```

Is there anything I can do on my end to help move this forward, or any interim workaround you’d recommend?

Thanks for your time and efforts! 🙏

timkobiolka commented:
Thanks for the feedback everyone! I'll have more time starting next week to update this merge request with the current main branch code.

aCodeDog commented:
I was wondering if this feature is still being worked on? I recently implemented the same feature [only modify raycaster.py without any loop during raycasting ]  and maybe I can contribute something to it. 
<img width="1665" height="934" alt="raycasting" src="https://github.com/user-attachments/assets/4e04470f-2180-4147-b950-2ce45c4ba18c" />

pascal-roth commented:
Thanks for the feature option, there is a new PR covering the multi-mesh integration with a more tuned implementation, closing this now in favor of #3298
